zerver/lib/user_topics.py

```diff
@@ -233,24 +233,26 @@ def mute_cond(row: RecipientTopicDict) -> ClauseElement:
     return [*conditions, condition]
 
 
-def build_topic_mute_checker(user_profile: UserProfile) -> Callable[[int, str], bool]:
-    rows = UserTopic.objects.filter(
-        user_profile=user_profile, visibility_policy=UserTopic.VisibilityPolicy.MUTED
-    ).values(
+def build_topic_visibility_policy_checker(
+    user_profile: UserProfile,
+) -> Callable[[int, str, int], bool]:
+    rows = UserTopic.objects.filter(user_profile=user_profile).values(
         "recipient_id",
         "topic_name",
+        "visibility_policy",
     )
```

Comment:
I think a single user doesn't have a very large number of user topic rows; so this shouldn't eat a lot of memory?

Or we should create three separate functions: `build_topic_mute_checker`, `build_topic_unmute_checker`, and `build_topic_follow_checker` ?

Reply:
Yeah, I think this should be fine; and it's a set data structure, so performance should be fine too.

Reply:
Thinking about this a bit more -- this still somewhat feels like the wrong API for how this is used, and since we only use it in one place, it might make sense to write it a bit more; with the same work, we could have this API be `get_prefetched_visibility_policy(recipient_id, topic) => visbiility_policy`, and call that function just once in `is_row_muted`; I don't love calling the same function 3 times with different parameters when we could just have the first line of `is_row_muted` be `visibility_policy = get_prefetched_visibility_policy_checker(recipient_id, topic)`

Since this is a loop through a very large number of messages, this sort of performance thing might matter.

Reply:
Makes sense.

Now, we have:
```
get_topic_visibility_policy = build_get_topic_visibility_policy(user_profile)

// within loop
visibility_policy = get_topic_visibility_policy(recipient_id, topic)
```
```
def build_get_topic_visibility_policy(
    user_profile: UserProfile,
) -> Callable[[int, str], int]:
    """Prefetch the visibility policies the user has configured for
    various topics.

    The prefetching helps to avoid the db queries later in the loop
    to determine the user's visibility policy for a topic.
    """
    rows = UserTopic.objects.filter(user_profile=user_profile).values(
        "recipient_id",
        "topic_name",
        "visibility_policy",
    )

    topic_to_visiblity_policy: Dict[Tuple[int, str], int] = defaultdict(int)
    for row in rows:
        recipient_id = row["recipient_id"]
        topic_name = row["topic_name"]
        visibility_policy = row["visibility_policy"]
        topic_to_visiblity_policy[(recipient_id, topic_name)] = visibility_policy

    def get_topic_visibility_policy(recipient_id: int, topic: str) -> int:
        return topic_to_visiblity_policy[(recipient_id, topic.lower())]

    return get_topic_visibility_policy
```